File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
def solution(n, wires):
    answer = -1
    visited = [False] * (n+1)
    
    lst = [[] for _ in range(n+1)]

    #dfs 탐색을 위해 새로운 리스트 만들기
    for i, j in wires:
        lst[i].append(j)
        lst[j].append(i)
        
    cnt = 0
    result = 0
    def dfs(cnt, visited, n):
        
        for i in lst[n]: #n에 연결된 전력망
            if visited[i] == False:
                visited[i] = True
                cnt += 1
                dfs(cnt, visited, i)
            
        
    #[1,2]를 끊는다고 하면 lst에서 lst[1] = 2를 삭제하고 탐색
    result = 100
    for n1, n2 in wires:
        lst[n1].remove(n2)
        lst[n2].remove(n1)
       
        visited = [False] * (n+1)
        visited[n1] = True
        dfs(1, visited, n1)
        cnt1 = visited.count(True)
        
        visited = [False] * (n+1)
        visited[n2] = True
        dfs(1, visited, n2)
        cnt2 = visited.count(True)
        
    
        result = min(result, abs(cnt1-cnt2))
        lst[n1].append(n2)
        lst[n2].append(n1)
            
    answer = result
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {

    public int solution(int n, int[][] wires) {
        int min = n;
        int answer = -1;
        
        for (int j = 0; j < wires.length; j++) {
            List<List<Integer>> list = new ArrayList<>();
            for(int i = 0; i < n; i++) {
                list.add(new ArrayList<>());
            }
            boolean[] visited = new boolean[n];
            
            for (int i = 0; i< wires.length; i++){
                //간선하나 제외하고 인접리스트 만들기
                if (i == j) continue; //j번째 간선 제거
                int a = wires[i][0] - 1;
                int b = wires[i][1] - 1;
                list.get(a).add(b);
                list.get(b).add(a);
            }
            visited[0] = true;
            int cnt = dfs(0, list, visited);
            min = Math.min(min, Math.abs(cnt - (n-cnt)));
        }
        return min;
    }
    private int dfs(int start, List<List<Integer>> list, boolean[] visited) {
        int cnt = 1;
        for (int next : list.get(start)) {
            if (!visited[next]) {
                visited[next] = true;
                cnt += dfs(next, list, visited);
            }
        }
        return cnt;
        
    }
    
    
}